rustls/src/conn/unbuffered.rs

```diff
@@ -0,0 +1,392 @@
+//! Unbuffered connection API
+
+use alloc::vec::Vec;
+use core::num::NonZeroUsize;
+use core::{fmt, mem};
+use std::error::Error as StdError;
+
+use crate::msgs::deframer::MessageDeframerSliceBuffer;
+use crate::{Error, UnbufferedConnectionCommon};
+
+impl<Data> UnbufferedConnectionCommon<Data> {
+    /// Processes TLS records in the `incoming_tls` buffer
+    pub fn process_tls_records<'c, 'i>(
+        &'c mut self,
+        incoming_tls: &'i mut [u8],
+    ) -> Result<UnbufferedStatus<'c, 'i, Data>, Error> {
+        let mut discard = 0;
+        let mut buffer = MessageDeframerSliceBuffer::new(incoming_tls, &mut discard);
+
+        let state = 'outer: loop {
+            if let Some(chunk) = self
+                .core
+                .common_state
+                .received_plaintext
+                .pop()
+            {
+                break 'outer AppDataAvailable::new(self, incoming_tls, chunk).into();
+            }
+
+            if let Some(chunk) = self
+                .core
+                .common_state
+                .sendable_tls
+                .pop()
+            {
+                break 'outer MustEncodeTlsData::new(self, chunk).into();
+            }
+
+            if let Some(msg) = self.core.deframe(&mut buffer)? {
+                let mut state =
+                    match mem::replace(&mut self.core.state, Err(Error::HandshakeNotComplete)) {
+                        Ok(state) => state,
+                        Err(e) => {
+                            self.core.state = Err(e.clone());
+                            return Err(e);
+                        }
+                    };
+
+                match self.core.process_msg(msg, state) {
+                    Ok(new) => state = new,
+
+                    Err(e) => {
+                        self.core.state = Err(e.clone());
+                        return Err(e);
+                    }
+                }
+
+                self.core.state = Ok(state);
+            } else if self.wants_write {
+                break MustTransmitTlsData { conn: self }.into();
```

Comment:
My idea (hope?) was to make it easier to batch records into a single TCP write by having an explicit `MustTransmitTlsData` state. without it, the end user may be tempted to do a `write` on each `MustEncode` state. As implemented, the examples certainly batch records so that only 1.5 - 2 (depending on the TLS version) round-trips are necessary to complete the handshake

I think we can remove this `wants_write` field and clearly document that after `MustEncode` has been handled the TLS must *eventually* be transmitted but not necessarily be transmitted *each time* `MustEncode` is handled because that results in no batching (more round-trips) -- that feels a bit hard to wrap the head around. I guess we could document that `outgoing_tls` must be transmitted in the `NeedsMoreTlsData` and `TrafficTransit` states if it's not empty ... would that cover all the cases :thinking:  ?

Reply:
Right, I think I understand and agree it is a necessity to have this to signal to consumers when we've finished writing a flight of messages; so they can flush/uncork/etc. the socket.